README.md
![django-leaf](https://s3.amazonaws.com/django-leaf/img/django-leaf-icon.png)

# django-leaf

> Render django templates based on URL path.

[![Build Status](https://img.shields.io/travis/gsmke/django-leaf/master.svg?style=flat)](https://travis-ci.org/gsmke/django-leaf)
[![Latest Version](https://img.shields.io/pypi/v/django-leaf.svg?style=flat)](https://pypi.python.org/pypi/django-leaf/)

## Quick start

1. Install the package from pypi:

    ```bash
    pip install django-leaf
    ```

2. Add "leaf" and "mptt" to your INSTALLED_APPS:

    ```python
    INSTALLED_APPS = (
        ...
        'leaf',
        'mptt',
    )
    ```

3. Add leaf urls to *the end* of your urlpatterns:

    ```python
    url(r'^', include('leaf.urls')),
    ```

## Usage

django-leaf can be used to render both "static" and database-backed templates.

### Static Templates

If you want to render a template when a user goes to `/example/url/`, create one of the following files:

1. example/url.html
2. example/url/index.html
3. pages/example/url.html
4. pages/example/url/index.html

### Database Backed Templates

After installing django-leaf, the admin interface will have a new section called `Pages` where you'll be able to create your page hierarchy.

To define your own page model, you need to extend from `leaf.models.Page`.
There are a few fields available for customization:

1. **identifier**: A unique identifier for your model. This will be used to associate page nodes with your page implementation. If you don't provide an `identifier`, one will be provided for you.
2. **template**: The template to render.
3. **admin_page_inline**: The admin class to use when rendering the template fields inline. This defaults to the default ``admin.StackedInline``.
4. **admin_inlines**: A list of other inlines to add to the admin.

Here's an example for creating a page with translations provided by [django-parler](https://github.com/edoburu/django-parler):

```python
# admin.py
from parler.admin import TranslatableStackedInline


class AboutPageInline(TranslatableStackedInline):
    pass

# models.py
from django.db import models
from leaf.models import Page
from parler.models import TranslatableModel, TranslatedFields

from .admin import AboutPageInline


class AboutPage(Page, TranslatableModel):
    admin_page_inline = AboutPageInline
    identifier = 'about-page'
    template = "about.html"

    translations = TranslatedFields(
        headline=models.CharField(max_length=255),
        copy=models.TextField(blank=True)
    )

    def __unicode__(self):
        return self.headline
```

When rendering the template, all of the model fields will be available on the ``page`` context variable:

```django
{{ page.headline }}
{{ page.copy }}
```

#### Home Page

The root page can be added to django-leaf by adding a new page with `slug=home` and `parent=None`. All pages added under that will be added without the ``home`` part in the path.

# TODO

1. Better documentation.
2. More configuration options.


CHANGELOG.md
# Changelog

<!---
Boilerplate:

## vX.X.X (YYYY-MM-DD)

### Added

### Deprecated

### Removed

### Fixed

### Security
-->
## v2.2.1 (2020-05-05)

- Add support for Django 2.2
- Update mptt dependency to mptt==0.11.0
- For anyone updating to this version, it requires a migration

## v2.2.0 (2015-08-24)

- Improve jinja2 support ([#2](https://github.com/gsmke/django-leaf/pull/2))
- Add Code of Conduct
- Add contributors attribution
- Add support for Django 1.10+ ([#3](https://github.com/gsmke/django-leaf/pull/3))

## v2.1.2 (2015-08-24)

- Fix version of mptt

## v2.1.1 (2015-05-28)

- Fix __str__ on Python 3

## v2.1.0 (2015-05-18)

- Add support for the home ('/') page.

## v2.0.0 (2015-05-15)

- Now supports adding additional inlines

## v1.1.0 (2015-04-29)

### Added

- Support for defining pages in the admin.

## v1.0.0 (2015-04-13)

Initial version.

### Added

- Support for rendering django templates based on url path


leaf/models.py
import six

from django.db import models
from django.utils.encoding import python_2_unicode_compatible
from mptt.models import MPTTModel, TreeForeignKey

PAGE_MODEL_CLASSES = []


class PageBase(models.base.ModelBase):

    """Automatiacally egister page models.

    This class is meant to be used as a __metaclass__:

    .. code-block:: python
        class Page(metaclass=Page):

    """

    def __init__(cls, name, bases, dct):
        super(PageBase, cls).__init__(name, bases, dct)

        # Dj < 1.10
        try:
            if cls._deferred:
                return
        # Dj >= 1.10
        except AttributeError:
            if cls is models.DEFERRED:
                return

        if not cls._meta.abstract:
            if not getattr(cls, 'identifier', None):
                cls.identifier = '{}.{}'.format(cls._meta.app_label, cls._meta.model_name)

            PAGE_MODEL_CLASSES.append(cls)


class Page(six.with_metaclass(PageBase, models.Model)):

    """Abstract base model for all leaf pages."""

    node = models.OneToOneField('PageNode', related_name='+', on_delete=models.CASCADE)

    class Meta:
        abstract = True


@python_2_unicode_compatible
class PageNode(MPTTModel):

    """Basic page node for all leaf pages."""

    slug = models.SlugField(max_length=50, unique=True)
    parent = TreeForeignKey('self', null=True, blank=True, related_name='children', db_index=True, on_delete=models.CASCADE)
    template = models.CharField(max_length=255, blank=True)
    path = models.CharField(max_length=255, blank=True, db_index=True)

    class MPTTMeta:
        order_insertion_by = ['slug']

    class Meta:
        unique_together = ('slug', 'parent')
        verbose_name = "page"

    def __str__(self):
        return self.slug

    def save(self, *args, **kwargs):
        if self.parent:
            if self.parent.path == 'home':
                # If the parent page is the home page, start
                # all paths from the root.
                self.path = '{}'.format(self.slug)
            else:
                self.path = '{}/{}'.format(self.parent.path, self.slug)
        else:
            self.path = self.slug

        super(PageNode, self).save(*args, **kwargs)
